core/src/price_estimation/priority_price_source.rs

```diff
@@ -0,0 +1,141 @@
+use super::PriceSource;
+use crate::models::TokenId;
+use anyhow::Result;
+use futures::future::{BoxFuture, FutureExt as _};
+use std::collections::{HashMap, HashSet};
+use std::iter::FromIterator;
+
+/**
+ * A price source that sequentially queries its inner sources in order and returns the
+ * first price found.
+ */
+pub struct PriorityPriceSource {
+    sources: Vec<Box<dyn PriceSource + Send + Sync>>,
+}
+
+impl PriorityPriceSource {
+    #[allow(dead_code)]
+    pub fn new(sources: Vec<Box<dyn PriceSource + Send + Sync>>) -> Self {
+        Self { sources }
+    }
+}
+
+impl PriceSource for PriorityPriceSource {
+    fn get_prices<'a>(
+        &'a self,
+        tokens: &'a [TokenId],
+    ) -> BoxFuture<'a, Result<HashMap<TokenId, u128>>> {
+        async move {
+            let mut remaining_tokens: HashSet<_> = HashSet::from_iter(tokens.iter().cloned());
+            let mut result = HashMap::new();
+            for source in &self.sources {
+                let remaining_token_vec = Vec::from_iter(remaining_tokens.iter().cloned());
+                match source.get_prices(&remaining_token_vec).await {
+                    Ok(partial_result) => {
+                        remaining_tokens = remaining_tokens
+                            .into_iter()
+                            .filter(|token| !partial_result.contains_key(token))
+                            .collect();
```

Comment:
Can probably be replaced with `HashSet::retain`. Or even make remaining tokens be `let mut remaining_tokens = tokens.to_vec()` then use `retain` on that and pass `&remaining_tokens` to `get_prices` since we aren't actually making use of this being a set instead of a vec.

Reply:
Right, O(1) lookup is only used for partial result so hash set is not needed. Will use a vec instead.

Reply:
instead of filter and collect we can still use `retain` on the vector here

Reply:
<a href="https://gitme.me/image?url=https%3A%2F%2Fmedia1.giphy.com%2Fmedia%2FkVVEP6SmfEXvy%2Fgiphy.gif&token=syntactic_sugar" data-gitmeme-token="syntactic_sugar"><img src="https://media1.giphy.com/media/kVVEP6SmfEXvy/giphy.gif" title="Created by gitme.me with /syntactic_sugar"/></a>